pkgs/test_core/lib/src/runner/reporter/github.dart

```diff
@@ -0,0 +1,198 @@
+
+    // Collect messages from tests as they are emitted.
+    _subscriptions.add(liveTest.onMessage.listen((message) {
+      if (_completedTests.contains(liveTest)) {
+        // The test has already completed and it's previous messages were
+        // written out; ensure this post-completion output is not lost.
+        _sink.writeln(message.text);
+      } else {
+        _testMessages.putIfAbsent(liveTest, () => []).add(message);
+      }
+    }));
+  }
+
+  /// A callback called when [liveTest] finishes running.
+  void _onComplete(LiveTest test) {
+    final errors = test.errors;
+    final messages = _testMessages[test] ?? [];
+    final skipped = test.state.result == Result.skipped;
+    final failed = errors.isNotEmpty;
+    final loadSuite = test.suite is LoadSuite;
+    final synthetic = loadSuite ||
+        test.individualName == setUpAllName ||
+        test.individualName == tearDownAllName;
+
+    // Mark this test as having completed.
+    _completedTests.add(test);
+
+    // Don't emit any info for loadSuite, setUpAll, or tearDownAll tests
+    // unless they contain errors or other info.
+    if (synthetic && (errors.isEmpty && messages.isEmpty)) {
+      return;
+    }
+
+    var defaultIcon =
+        synthetic ? _GithubHelper.synthetic : _GithubHelper.passed;
+    final prefix = failed
+        ? _GithubHelper.failed
+        : skipped
+            ? _GithubHelper.skipped
+            : defaultIcon;
+    final statusSuffix = failed
+        ? ' (failed)'
+        : skipped
+            ? ' (skipped)'
+            : '';
+
+    var name = test.test.name;
+    if (!loadSuite) {
+      if (_printPath && test.suite.path != null) {
+        name = '${test.suite.path}: $name';
+      }
+    }
+    _sink.writeln(_helper.startGroup('$prefix $name$statusSuffix'));
+    // TODO: Note that printing messages and errors in this manner could display
+    // them in a different order than they were generated.
+    for (var message in messages) {
+      _sink.writeln(message.text);
+    }
+    for (var error in errors) {
+      _sink.writeln('${error.error}');
+      _sink.writeln(error.stackTrace.toString().trimRight());
+    }
+    _sink.writeln(_helper.endGroup);
+  }
+
+  void _onDone(bool? success) {
+    _cancel();
+
+    _sink.writeln();
+
+    final hadFailures = _engine.failed.isNotEmpty;
+    final message = StringBuffer('${_engine.passed.length} '
+        '${pluralize('test', _engine.passed.length)} passed');
+    if (_engine.failed.isNotEmpty) {
+      message.write(', ${_engine.failed.length} failed');
+    }
+    if (_engine.skipped.isNotEmpty) {
+      message.write(', ${_engine.skipped.length} skipped');
+    }
+    message.write('.');
+    _sink.writeln(
+      hadFailures
+          ? _helper.error(message.toString())
+          : '${_GithubHelper.success} $message',
+    );
+  }
+}
+
+class _GithubHelper {
+  // Char sets avilable at https://www.compart.com/en/unicode/.
+  static const String passed = '✅';
```

Comment:
[nit] I have a slight preference for `'\u2705'`, but I'm also curious what other folks think.

@grouma @jakemac53 @lrhn ?

Reply:
`'\u2705'` is NOT super readable 🤷  – why the preference?

Reply:
Historically they've not worked well in some editors and they aren't easy to type.
Mostly we just haven't used them in source strings that I recall, and I've only seen the `\u` version used.

Reply:
I changed these to use the unicode escaping form.

Reply:
Sorry for the late response. I think I prefer `\u2705` as a constant variable with an appropriate name, e.g. `_checkMark`. That way it's readable and will work with all tools / editors.

Reply:
One point against the `'\u'` form is that the Java and JavaScript style guides at google both seem to encourage it when it's "easier to read", which I think does apply here.
https://google.github.io/styleguide/javaguide.html#s2.3.3-non-ascii-characters
https://google.github.io/styleguide/jsguide.html#non-ascii-characters

C++ mentions that no user facing strings should be in source, which isn't really a practice for us, but they do still specifically allow non-ascii in source.
https://google.github.io/styleguide/cppguide.html#Non-ASCII_Characters

I'm very willing to accept that my preference is old-fashioned and choose a more modern style for our source code if that's the consensus - I mostly just wanted to double check that we weren't introducing a new pattern that would become standard without thinking it through.

@munificent - do you think this is something we should cover in Effective Dart?

Reply:
Fwiw for me having the actual characters is better, I have no idea what `\u2705` is and would have to look it up. I don't really see any downside to using the actual character and you have the advantage of seeing it visually.

Reply:
> @munificent - do you think this is something we should cover in Effective Dart?

Seems pretty nit-picky to me, but if you think it's worth it, I'm persuadable.

Reply:
I think - for readability of this particular piece of code - I'm going to switch back to having the actual char in the string. It's not causing issues w/ tooling, and makes it much easier to read this bit od code.

Reply:
FWIW, I'm generally comfortable with having non-ASCII characters inside Dart comments and string literals. It's entirely well-specified by the language and there's nothing intrinsically better about Unicode codepoints whose value happens to be less than 128. (I, of course, avoid homoglyphs and other nasty ones.)